Source/Core/Core/HW/EXI/BBA/BuiltIn.cpp

```diff
@@ -0,0 +1,705 @@
+      *ipdata = Common::IPv4Header((u16)(datasize + 8), IPPROTO_UDP, ref->from, ref->to);
+
+      *udpdata = Common::UDPHeader(ref->from, ref->to, (u16)datasize);
+      udpdata->checksum = Common::ComputeTCPNetworkChecksum(ref->from, ref->to, udpdata,
+                                                            (u16)(datasize + 8), IPPROTO_UDP);
+      datasize += 0x2a;
+    }
+    break;
+
+  case IPPROTO_TCP:
+    sf::Socket::Status st = sf::Socket::Status::Done;
+    TcpBuffer* tcp_buffer = nullptr;
+    for (auto& tcp_buf : ref->tcp_buffers)
+    {
+      if (tcp_buf.used)
+        continue;
+      tcp_buffer = &tcp_buf;
+      break;
+    }
+
+    // set default size to 0 to avoid issue
+    datasize = 0;
+    const bool can_go = (GetTickCountStd() - ref->poke_time > 100 || ref->window_size > 2000);
+    if (tcp_buffer != nullptr && ref->ready && can_go)
+      st = ref->tcp_socket.receive(&buffer[0x36], 440, datasize);
+
+    if (datasize > 0)
+    {
+      std::memset(buffer, 0, 0x36);
+      auto [hwpart, ipdata, tcpdata] = GetTCPHeaders(buffer, ref->bba_mac, ref->my_mac);
+
+      *ipdata = Common::IPv4Header((u16)(datasize + 20), IPPROTO_TCP, ref->from, ref->to);
+
+      *tcpdata = Common::TCPHeader(ref->from, ref->to, ref->seq_num, ref->ack_num, TCP_FLAG_ACK);
+      tcpdata->checksum = Common::ComputeTCPNetworkChecksum(ref->from, ref->to, tcpdata,
+                                                            (u16)(datasize + 20), IPPROTO_TCP);
+
+      // build buffer
+      tcp_buffer->seq_id = ref->seq_num;
+      tcp_buffer->data_size = (u16)datasize + 0x36;
+      tcp_buffer->tick = GetTickCountStd();
+      std::memcpy(&tcp_buffer->data[0], buffer, datasize + 0x36);
+      tcp_buffer->seq_id = ref->seq_num;
+      tcp_buffer->used = true;
+      ref->seq_num += (u32)datasize;
+      ref->poke_time = GetTickCountStd();
+      datasize += 0x36;
+    }
+    if (GetTickCountStd() - ref->delay > 3000)
+    {
+      if (st == sf::Socket::Disconnected || st == sf::Socket::Error)
+      {
+        ref->ip = 0;
+        ref->tcp_socket.disconnect();
+        datasize = BuildFINFrame(ref, buffer);
+      }
+    }
+    break;
+  }
+
+  return datasize;
+}
+
+void CEXIETHERNET::BuiltInBBAInterface::ReadThreadHandler(CEXIETHERNET::BuiltInBBAInterface* self)
+{
+  while (!self->m_read_thread_shutdown.IsSet())
+  {
+    // make thread less cpu hungry
+    std::this_thread::sleep_for(std::chrono::milliseconds(1));
+
+    if (!self->m_read_enabled.IsSet())
+      continue;
+    size_t datasize = 0;
+
+    u8 wp = self->m_eth_ref->page_ptr(BBA_RWP);
+    const u8 rp = self->m_eth_ref->page_ptr(BBA_RRP);
+    if (rp > wp)
+      wp += 16;
+
+    if ((wp - rp) >= 8)
+      continue;
+
+    std::lock_guard<std::mutex> lock(self->m_mtx);
+    // process queue file first
+    if (self->m_queue_read != self->m_queue_write)
+    {
+      datasize = self->m_queue_data[self->m_queue_read].size();
```

Comment:
AFAICT, `mRecvBuffer` size is `BBA_RECV_SIZE` and datasize isn't checked against it.